grakn-graph/src/test/java/ai/grakn/graph/property/EntityTypePropertyTest.java

```diff
@@ -0,0 +1,163 @@
+ * Copyright (C) 2016  Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ *
+ */
+
+package ai.grakn.graph.property;
+
+import ai.grakn.GraknGraph;
+import ai.grakn.concept.Entity;
+import ai.grakn.concept.EntityType;
+import ai.grakn.exception.ConceptException;
+import ai.grakn.generator.AbstractTypeGenerator.Meta;
+import ai.grakn.generator.FromGraphGenerator.FromGraph;
+import ai.grakn.generator.GraknGraphs.Open;
+import com.pholser.junit.quickcheck.Property;
+import com.pholser.junit.quickcheck.runner.JUnitQuickcheck;
+import org.junit.Rule;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+
+import static ai.grakn.graph.property.PropertyUtil.choose;
+import static ai.grakn.graph.property.PropertyUtil.directSubTypes;
+import static ai.grakn.util.ErrorMessage.META_TYPE_IMMUTABLE;
+import static ai.grakn.util.ErrorMessage.SUPER_TYPE_LOOP_DETECTED;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.not;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assume.assumeThat;
+
+/**
+ * @author Felix Chapman
+ */
+@RunWith(JUnitQuickcheck.class)
+public class EntityTypePropertyTest {
+
+    @Rule
+    public ExpectedException exception = ExpectedException.none();
+
+    @Property
+    public void whenDeletingMetaEntityType_Throw(@Open GraknGraph graph) {
+        EntityType entity = graph.admin().getMetaEntityType();
+
+        exception.expect(ConceptException.class);
+        exception.expectMessage(META_TYPE_IMMUTABLE.getMessage(entity.getName()));
+
+        entity.delete();
+    }
+
+    @Property
+    public void whenANonMetaEntityTypeHasNoInstancesSubTypesOrRules_ItCanBeDeleted(
+            @Open GraknGraph graph, @FromGraph @Meta(false) EntityType type) {
+        assumeThat(type.instances(), empty());
+        assumeThat(type.subTypes(), contains(type));
+        assumeThat(type.getRulesOfHypothesis(), empty());
+        assumeThat(type.getRulesOfConclusion(), empty());
+
+        type.delete();
+
+        assertNull(graph.getType(type.getName()));
+    }
+
+    @Property
+    public void whenAddingAnEntityOfAMetaType_Throw(@Meta EntityType type) {
```

Comment:
Does this take into account the rule meta types `MetaSchema.INFERENCE_RULE` and `MetaSchema.CONSTRAINT_RULE` ? You allowed to add instances to those.